### PR TITLE
Deploy the whole jitsi stack on Kubernetes

### PR DESCRIPTION
**Pull request by madmatah**
## Purpose

Now that we have a working Jibri setup deployed on Kubernetes, we want to depoly the entire jitsi-meet stack on our Kubernetes cluster.

## Proposal

- [x] Re-organize the project structure using Kustomize overlays
- [x] Deploy prosody
- [x] Deploy jicofo
- [x] Deploy jitsi-meet frontend
- [x] Deploy jitsi videobridge
- [x] Update documentation
- [x] Add tool to setup a new environment / overlay
- [x] Allow to change the overlay directory (to put user specific overlays in a separate repository)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,7 @@ env.d/*.env-*
 
 # Generated Kubernetes configuration file
 .kubeconfig*
+
+k8s/overlays/*
+!k8s/overlays/example
+!k8s/overlays/.template
```
Comment (sampaccoud) [inline .gitignore:14]: I didn't know you can do that. You could also have added the 2 files explicitly.

File: README.md [hunk 2]
```diff
@@ -18,7 +18,7 @@ This repository is composed of two parts:
    the Kubernetes cluster.
 
 
-We made the following (opinionated) choices, but the project can be easily adapted to choose alternatives options :
+We made the following (opinionated) choices, but the project can be easily adapted to choose alternatives options:
```
Comment (sampaccoud) [inline README.md:21]:
```suggestion
We made the following (opinionated) choices, but the project can be easily adapted to choose alternative options:
```

File: env.d/kustomize.dist [hunk 3]
```diff
@@ -0,0 +1,7 @@
+
+# Path to the kustomize overlays.
+# If you want store your overlays on a separate directory,
```
Comment (sampaccoud) [inline env.d/kustomize.dist:3]:
```suggestion
# If you want to store your overlays in another directory,
```

File: k8s/base/jibri-hpa.yaml [hunk 4]
```diff
@@ -56,7 +56,7 @@ spec:
         describedObject:
           apiVersion: v1
           kind: Namespace
-          name: jibri
+          name: jitsi
```
Comment (sampaccoud) [inline k8s/base/jibri-hpa.yaml:59]: the filename mentions jibri. Is it a fail?

File: k8s/base/jitsi-meet-front-ingress.yml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+kind: Ingress
+metadata:
+  name: "jitsi-meet-front"
+  labels:
+    app: "jitsi-meet-front"
+  annotations:
+    cert-manager.io/issuer: "letsencrypt-jitsi-issuer"
+spec:
+  rules:
+  - host: "placeholder.domain.example"
```
Comment (sampaccoud) [inline k8s/base/jitsi-meet-front-ingress.yml:14]: make it a variable?

File: .gitignore (hunk 1 above)
Comment (madmatah) [reply .gitignore:14]: We are using this kind of rules in [other](https://github.com/openfun/marsha/blob/master/.gitignore#L32) [projects](https://github.com/openfun/richie/blob/master/.gitignore#L97).
I like when rules are explicit and match with the reality :grin:

Comment (sampaccoud) [reply .gitignore:14]: My bad, I thought you were French!

File: k8s/base/jibri-hpa.yaml (hunk 4 above)
Comment (madmatah) [reply k8s/base/jibri-hpa.yaml:59]: No it's not.
The goal of this `DescribedObject` section is to describe to which object this metric corresponds.
And in this case, this is a metric that correspond to the entire namespace in which jibri is deployed.

The namespace has been renamed to `jitsi` (it was `jibri` before)

File: k8s/base/jitsi-meet-front-ingress.yml (hunk 5 above)
Comment (madmatah) [reply k8s/base/jitsi-meet-front-ingress.yml:14]: Do you have a specific suggestion on how to do this?

This placeholder value is not used: it is replaced by the content of the `XMPP_DOMAIN` variable with the replacement feature of kustomize, as described in the comment on top of this file.

File: k8s/base/prosody-deployment.yml [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      labels:
+        app: prosody
+    spec:
+      containers:
+      - image: jitsi/prosody:stable-5963
+        name: prosody
+        imagePullPolicy: IfNotPresent
+        livenessProbe:
+          httpGet:
+            path: /http-bind
```
Comment (lunika) [inline k8s/base/prosody-deployment.yml:24]: Does this path is available is you enable only websocket and not bosh ?

File: k8s/base/cert-issuer.yml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+apiVersion: cert-manager.io/v1
```
Comment (lunika) [inline k8s/base/cert-issuer.yml:1]: In you commit message a link seems missing. (`jitsi[1]` and nothing related to `1`)

**Review by lunika — COMMENTED** [review]
Great improvement ! I just don't understand how JVB nodepools are reserved to JVB resources. Does a prosody or jicofo pod can be deployed on them ?

File: k8s/base/prosody-deployment.yml (hunk 6 above)
Comment (madmatah) [reply k8s/base/prosody-deployment.yml:24]: Good question! I just checked:
In the configuration file provided with the `jitsi/prosody` docker image, bosh is always enabled, so this endpoint is always available.

See https://github.com/jitsi/docker-jitsi-meet/blob/master/prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua#L110-L112

**Comment by madmatah** [comment]
> Great improvement ! I just don't understand how JVB nodepools are reserved to JVB resources. Does a prosody or jicofo pod can be deployed on them ?

Thanks!
Each deployment has a nodeselector specified, to restrict the scheduling of the pods on the right nodepool.
It uses labels provided by Scaleway.

Example for prosody:
```
      nodeSelector:
        k8s.scaleway.com/pool-name: default
```

And for JVBs:
```
      nodeSelector:
        k8s.scaleway.com/pool-name: jibri
```

This was the simplest solution to implement.

File: k8s/base/jvb-statefulset.yml [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+              path: /about/health
+              port: 8080
+          readinessProbe:
+            httpGet:
+              path: /about/health
+              port: 8080
+          resources:
+            {}
+      nodeSelector:
+        k8s.scaleway.com/pool-name: jibri
```
Comment (madmatah) [inline k8s/base/jvb-statefulset.yml:78]:
```suggestion
        k8s.scaleway.com/pool-name: jvb
```

File: k8s/base/jibri-hpa.yaml (hunk 4 above)
Comment (sampaccoud) [reply k8s/base/jibri-hpa.yaml:59]: Ok that's what I missed :+1:

File: k8s/base/jitsi-meet-front-ingress.yml (hunk 5 above)
Comment (sampaccoud) [reply k8s/base/jitsi-meet-front-ingress.yml:14]: Yes that's the part I'm not confortable with as we can end-up with example settings in production right?

**Review by lunika — APPROVED** [review]

**Review by sampaccoud — APPROVED** [review]

**Pull request merged**